x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,341 @@
+        updateMapping(imdVersion -> {
+            currentIndexMetadataVersion = imdVersion;
+            LOGGER.info("{} Started to follow leader shard {}, followGlobalCheckPoint={}, indexMetaDataVersion={}",
+                params.getFollowShardId(), params.getLeaderShardId(), followGlobalCheckpoint, imdVersion);
+            coordinateReads();
+        });
     }
 
+    private synchronized void coordinateReads() {
+        if (isStopped()) {
+            LOGGER.info("{} shard follow task has been stopped", params.getFollowShardId());
+            return;
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final int maxReadSize = params.getMaxBatchOperationCount();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, from, maxReadSize);
+            long maxRequiredSeqno = Math.min(globalCheckpoint, from + maxReadSize);
+            sendShardChangesRequest(from, maxReadSize, maxRequiredSeqno);
+            lastRequestedSeqno = maxRequiredSeqno;
+        }
+
+        if (numConcurrentReads == 0) {
+            // We sneak peek if there is any thing new in the leader.
+            // If there is we will happily accept
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, from);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReadBatches()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxWriteBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            long sumEstimatedSize = 0L;
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxBatchOperationCount(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+                sumEstimatedSize += ops[i].estimateSize();
+                if (sumEstimatedSize > params.getMaxBatchSizeInBytes()) {
+                    ops = Arrays.copyOf(ops, i + 1);
```

Comment:
I wonder if we should use an ArrayList with initial capacity. We can then change the request etc to use List<> instead of array

Reply:
In that case we always make a copy of the underlying array in ArrayList, while in this case we only make a copy if maxBatchSizeInBytes limit has been reached.

Reply:
> In that case we always make a copy of the underlying array in ArrayList,

Not if you change the request and such to use lists.

> while in this case we only make a copy if maxBatchSizeInBytes limit has been reached.
which may be frequent (i.e., every request), which is why I was considering making a change.